_posts/2018-10-03-ssl-logging-in-otp-22.md

```diff
@@ -0,0 +1,270 @@
+messages.
+
+## Implementation
+
+In the ```ssl``` application, we needed a way to handle two types of protocol
+messages, tls_record and handshake, each with a custom formatter.
+
+The most straightforward solution was to add a new handler instance to the logger with a
+special formatter function that filters out all the "noise" coming from other
+modules of the system.
+
+The handler itself could reuse the standard handler for logger, ```logger_std_h```, as it
+could print logs to ```standard_io```. You can add multiple standard handler instances to
+logger if your application requires it.
+
+```erlang
+logger:add_handler(ssl_handler, logger_std_h, Config),
+```
+
+The new ssl_handler is configured with a formatter that is implemented by the ```ssl_logger```
+module.
+
+```erlang
+Config = #{level => debug,
+           filter_default => stop,
+           formatter => {ssl_logger, #{}}},
+```
+
+Handler filter level is set to ```debug``` with ```stop``` as the default filter action. We also
+need a filter that lets the log events pass to the formatter if the source of the log event is the
+ssl application. In other words, we need a domain filter with the action ```log``` on all sub-domains
+matching ```[otp,ssl]```.
+
+```erlang
+Filter = {fun logger_filters:domain/2,{log,sub,[otp,ssl]}},
+```
+
+Putting it all together we get the following function.
+
+```erlang
+start_logger() ->
+    Config = #{level => debug,
+               filter_default => stop,
+               formatter => {ssl_logger, #{}}},
+    Filter = {fun logger_filters:domain/2,{log,sub,[otp,ssl]}},
+    logger:add_handler(ssl_handler, logger_std_h, Config),
+    logger:add_handler_filter(ssl_handler, filter_non_ssl, Filter).
+```
+
+The function ```format``` is called in ssl_logger when an event gets through all the filters:
+
+```erlang
+format(#{level:= _Level, msg:= {report, Msg}, meta:= _Meta},
+       _Config0) ->
+     #{direction := Direction,
+       protocol := Protocol,
+       message := BinMsg0} = Msg,
+    case Protocol of
+        'tls_record' ->
+            BinMsg = lists:flatten(BinMsg0),
+            format_tls_record(Direction, BinMsg);
+        'handshake' ->
+            format_handshake(Direction, BinMsg0);
+        _Other ->
+            []
+    end.
+```
+
+There are two more helper functions that wrap around the logging macros. They were added in order
+to be able to set logging level per TLS session.
+
+```erlang
+debug(Level, Report, Meta) ->
+    case logger:compare_levels(Level, debug) of
+        lt ->
+            ?LOG_DEBUG(Report, Meta);
+        eq ->
+            ?LOG_DEBUG(Report, Meta);
+        _ ->
+            ok
+    end.
+
+notice(Level, Report) ->
+    case logger:compare_levels(Level, notice) of
+        lt ->
+            ?LOG_NOTICE(Report);
+        eq ->
+            ?LOG_NOTICE(Report);
+        _ ->
+            ok
+    end.
+```
+
+Finally, the logger can be called with the debug level and domain parameters.
```

Comment:
I got a bit confused by this sentence - mostly since you say that "the logger can be called", while the example really calls ssl_logger. Maybe it would be clearer if you state something like "To print a log event, the above functions are called with the configured ssl log level and the domain parameter".

Reply:
Will fix.